Another builder error when using a sum type

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Pop!_OS 22.04 LTS
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i7-5600U CPU @ 2.60GHz
CC version: cc (Ubuntu 11.2.0-19ubuntu1) 11.2.0

getwd: /home/hedo/usr/tmp/v
vmodules: /home/hedo/.vmodules
vroot: /home/hedo/usr/tmp/v/v
vexe: /home/hedo/usr/tmp/v/v/v
vexe mtime: 2022-09-11 18:48:53
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.1 4bd49a0.550b27b

Git version: git version 2.34.1
Git vroot status: weekly.2022.36-29-g550b27b0
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 827f7452

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg problem.v`
```v
type Expr = Fun | int

struct Fun {
	f fn (int) int
}

fn main() {
	f := Expr(0)
	if f is Fun {
		println((f as Fun).f(0))
	}
}

```

**What did you expect to see?**

not a build error


**What did you see instead?**
```
==================
/tmp/v_1000/problem.2887101826652499131.tmp.c:12112: error: field not found: (
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1342,7 +1342,16 @@ fn (mut g Gen) fn_call(node ast.CallExpr) {
 			if obj := node.scope.find(node.name) {
 				match obj {
 					ast.Var {
-						if obj.smartcasts.len > 0 {
+						// Temp fix generate call fn error when the struct type of sumtype
+						// has the fn field and is same to the struct name.
+						mut is_need_cast := true
+						if node.left_type != 0 {
+							left_sym := g.table.sym(node.left_type)
+							if left_sym.kind == .struct_ && node.name == obj.name {
+								is_need_cast = false
+							}
+						}
+						if obj.smartcasts.len > 0 && is_need_cast {
 							for _ in obj.smartcasts {
 								g.write('(*')
 							}

--- a/vlib/v/tests/sumtype_with_struct_fn_field_call_test.v
+++ b/vlib/v/tests/sumtype_with_struct_fn_field_call_test.v
@@ -1,0 +1,13 @@
+type Expr = Fun | int
+
+struct Fun {
+	f fn (int) int
+}
+
+fn test_sumtype_with_struct_fn_field_call() {
+	f := Expr(0)
+	if f is Fun {
+		println((f as Fun).f(0))
+	}
+	assert true
+}

